napari/layers/labels/labels.py

```diff
@@ -1225,6 +1232,54 @@ def fill(self, coord, new_label, refresh=True):
         if refresh is True:
             self.refresh()
 
+    def _draw(self, new_label, last_cursor_coord, coordinates):
+        """Paint new label into layer at given coordinates,
+        interpolating from the last cursor coordinate.
+
+        Parameters
+        ----------
+        new_label : int
+            value of label to paint
+        last_cursor_coord : sequence
+            last painted cursor coordinates
+        coordinates : sequence
+            new cursor coordinates
+        """
+        if coordinates is None:
+            return
+        ndisplay = len(self._dims_displayed)
+        interp_coord = interpolate_coordinates(
+            last_cursor_coord, coordinates, self.brush_size
+        )
+        for c in interp_coord:
+            if (
+                ndisplay == 3
+                and self.data[tuple(np.round(c).astype(int))] == 0
+            ):
+                continue
+            if self._mode in [Mode.PAINT, Mode.ERASE]:
+                self.paint(c, new_label, refresh=False)
+            elif self._mode == Mode.FILL:
+                self.fill(c, new_label, refresh=False)
+        self.refresh()
+
+    @contextmanager
+    def block_history(self):
+        prev = self._block_history
+        self._block_history = True
+        try:
+            yield
+        finally:
+            self._finish_painting()
```

Comment:
@Czaki I've implemented with `try: finally:` here, does this look right? I'm not sure if `_finish_painting` should be called in the `finally` block or after the `yield` in the `try`...